draft-ietf-tls-esni.md

```diff
@@ -482,6 +482,48 @@ The decompression process in {{encoding-inner}} forbids
 "encrypted_client_hello" in OuterExtensions. This ensures the unauthenticated
 portion of ClientHelloOuter is not incorporated into ClientHelloInner.
 
+# The Handshake Padding Message {#padding-message}
+
+Encrypting the ClientHelloInner may not be sufficient to protect sensitive
+handshake parameters against traffic analysis. In particular, the length of the
+ClientHelloInner, or the length of the backend server's response, may leak
+information about these parameters. To mitigate this issue, the client
+implements the padding scheme specified in {{client-padding}}. The backend
+server uses the more general-purpose handshake-padding mechanism described in
+this section. (See {{backend-server-padding}} for a recommendation of how to
+choose the padding length.)
+
+A new handshake message type is defined, which a client and server can use to
+insert padding into the encrypted portion of the handshake.
+
+~~~~
+    enum {
+        padding(16),
+    } HandshakeType;
+~~~~
+
+[[TODO: Address IANA considerations for new codepoint.]] The message is
+structured as follows:
+
+~~~~
+    struct {
+        uint8 zeros[length_of_padding];
+    } Padding;
+~~~~
+
+where `zeros` consists of any number of zero bytes. A host that receives a
+Padding message MAY verify that the payload is correct and abort with
+"illegal_parameter" if not. A Padding message MAY be sent by the client or
```

Comment:
Let's not have optional messages. They complicate the handshake state machine and are a source of security bugs.

Instead, define when the message is sent (probably in ClientHelloInner handshakes where the server sends an confirmation signal). We need to do anyway because messages only work as an extension when point solicited. I.e. you need a signal from the peer that they're allowed before you send them. Once we've done that, we can include it unconditionally.

Reply:
Just so I understand this correctly: The client sends an empty extension, which the server may reply to with the Padding message. Is that right? Do we also want to allow the server to solicit the Padding message? E.g., if it sends the extension in EE, then the client may reply with a Padding message?

Reply:
(Good catch, btw. Not sure how I missed this!)

Reply:
That works mechanically, but I think it's not the right design. It still involves an optional message, which complicates the receiver's state machine. It also means servers must account for clients that offer ECH without this new extension.

Rather, we can say:

* If the client sends "encrypted_client_hello" with type "inner" AND the server sends the ECH confirmation signal, the connection has the "ECH confirmed" bit.
* Both client and server will agree on this bit and, if set, we know both clients and servers are new enough to know about ECH. That clears the compatibility bar for tweaking the handshake.
* If the connection has the "ECH confirmed" bit, the client and server MUST send and consume _exactly_ one Padding message before each Finished message. Making the message required means the state machine always knows the next message to expect.
* How many bytes go in the Padding message is up to the respective sender. If you don't have anything to pad, just send a zero-length message.
* If we decide we need in-handshake padding for other reasons, we can add some OR clauses to the second condition.

@martinthomson, I think this was also the construction you preferred?

Reply:
That's pretty much it.

I also like the option for a message after the handshake.  That doesn't go in the transcript.  @davidben can just hold off on sending client Finished until it is all received, to avoid his fun deadlocks.

Reply:
I think this is too clever to the point of demonstrating why post-handshake doesn't work. :-)

If the receiver does not block the handshake on it, I can tell whether it was padding by delaying that record and seeing if the connection is stuck. If the receiver does block the handshake on it, TCP concerns no longer apply and it's not really "after the handshake".

Moreover, making it post-handshake means it's after the key change, which gives a record boundary. Record boundaries are public, which means the padding becomes useless. It needs to be at handshake encryption level and buffered up and flushed to the record layer with the rest of the handshake flight.

Post-Finished doesn't even keep it out of the transcript because the transcript is {..., ServerCertificate, ServerCertificateVerify, ServerFinished, ClientCertificate, ClientCertificateVerify, ClientFinished}. Any server message after ServerFinished will naturally sandwich in there, unless we special case the message... at which point, why are we cleverly positioning it?

If we want it to be post-Finished but pre-key-change and handshake-blocking, that's perfectly well-defined, I suppose, but having "Finished" not be the message which "finishes" the handshake seems weird. :-) We also need it after CertificateVerify due to variable-length ECDSA signatures. Put those together and we've uniquely identified the position: in between CertificateVerify and Finished.

Reply:
@martinthomson and @davidben, I've rebased this PR and updated it with David's construction. Please have another look.